refactor(validate): Use `nvim_echo` for emitting config errors

OXY2DEV opened the pull request:
## Issue

The default error message is way too big and it's kinda hard to find the *actual* error part.

## Solution

Use `nvim_echo()` instead of `error()`.

> As this will most likely be run only once we shouldn't need `notify_once()`.

## Images

A few images of the new error is given below,

![Screenshot_2025-03-28-14-52-22-110_com termux-edit](https://github.com/user-attachments/assets/bd49e514-4ba9-4093-8186-4a1dfd501695)
<sup>Uses `nvim_echo()`</sup>
![Screenshot_2025-03-28-14-52-45-957_com termux-edit](https://github.com/user-attachments/assets/acdebd73-bb1d-4435-b697-7d1ae1290933)
<sup>Fallback to `vim.notify()` when the other function isn't available</sup>

Ref: #973

OXY2DEV commented:
I can't believe I am being nagged by `stylua` for using `;` of all things.

😭

Saghen commented:
This looks really cool! My only worry is that noice.nvim seems to swallow it on startup by default and the user has to do `:messages` to see the message, which they likely don't know exists

OXY2DEV commented:
@Saghen I think we can do something like,

```lua
if package.loaded["noice"] then
  vim.warn(msg);
end
```

I will probably play around a bit and see what I can do.

---

Also, should I rework the other `error()`s too, just for consistency?

I think some of the errors can be shortened a bit(to prevent hit-enter on Startup).

OXY2DEV commented:
So, I am checking around and this seems to be an issue with `vim.ui_attach()` itself.

I believe when the message is sent Neovim's UI isn't ready yet so noice can't show it.

But, I am checking and `nvim_echo()` & `notify()` get ignored.

I will do some more digging and see if I can do something.

Saghen commented:
Perhaps there's an autocmd we need to wait for like `UIEnter`?

Saghen commented:
> Also, should I rework the other `error()`s too, just for consistency?

I don't think we should remove the `error()` calls anywhere, including in the validate, since we still want to stop execution. But I'm all for doing the `nvim_echo` in the downloader code as well

OXY2DEV commented:
> Perhaps there's an autocmd we need to wait for like `UIEnter`?

I don't want to use autocmds as in certain cases Neovim will straight up ignore UI updates anyway(e.g. when confirmation window pops up, or in 0.10 when `:s/` was used). So, it wouldn't really be a better solution than just simply delaying the message.

I think I need to rework the logic for detecting when custom UIs are used.

OXY2DEV commented:
> I'm all for doing the nvim_echo in the downloader code as well. This would result in the error message appearing twice though

I will check if I can do anything about it.

OXY2DEV commented:
> > Also, should I rework the other `error()`s too, just for consistency?
> 
> I don't think we should remove the `error()` calls anywhere, including in the validate, since we still want to stop execution. But I'm all for doing the `nvim_echo` in the downloader code as well. This would result in the error message appearing twice though

Do I change all the errors in `lua/blink/cmp/fuzzy/download/init.lua`?

OXY2DEV commented:
Hey @Saghen, does this message look clear enough?

![Screenshot_2025-03-29-08-32-19-962_com termux-edit](https://github.com/user-attachments/assets/e29d0d2b-903e-40f0-9666-34d6512cd6fa)

I thought I use `fuzzy.prebuilt_binaries.{download}` instead of `fuzzy.prebuilt_binaries.download` as when I started it was a bit confusing to see the entire path of the option.

> I am using `vimdoc`'s parameter syntax which uses `{param}?` format.

This is what it will normally look like,
![Screenshot_2025-03-29-08-34-41-532_com termux-edit](https://github.com/user-attachments/assets/a483112f-2598-431a-9057-e6c5a706237c)

OXY2DEV commented:
@Saghen I have finished replacing the messages.

Some of them look as follows,
![Screenshot_2025-03-29-08-10-10-385_com termux-edit](https://github.com/user-attachments/assets/18426cf7-8468-4882-9eb5-f7f523be812d)
![Screenshot_2025-03-29-08-21-30-667_com termux-edit](https://github.com/user-attachments/assets/ad97c89e-5971-4d3c-881a-ded101dc6a78)
![Screenshot_2025-03-29-08-34-41-532_com termux-edit](https://github.com/user-attachments/assets/237dfe57-bda5-4867-adfd-e9d5a615d285)
![Screenshot_2025-03-29-10-10-50-044_com termux-edit](https://github.com/user-attachments/assets/2aa31e77-fbe1-4f62-981b-0db8428cec86)
![Screenshot_2025-03-29-10-59-33-760_com termux-edit](https://github.com/user-attachments/assets/27dc297a-79aa-468a-9ced-dbfacbd74fdc)

>[!NOTE]
> I have removed instances of `See the docs` as it's a bit ambiguous as the docs can be either the vimdoc help file, the README.md or the plugin's website.

>[!IMPORTANT]
> I have removed `= true` parts of certain message as it should be quite clear what "enabling" an option means.

I have changed the wording of the messages so you can review them, if you have the time.

>[!BUG]
> There's a bug with `noice` related to highlight groups as noice appears to *replace* the messages highlight groups with it's own highlight groups. So, the first printed message may sometimes have incorrect highlight groups.

There's also another issue with `noice` consuming multiple messages at startup and only showing the last one. But as far as I am aware this is currently a limitation of `vim.ui_attach()` so there's not much we can do about it(delaying the message doesn't do anything).

Feel free to propose changes.

Saghen commented:
It seems like `nvim_echo` is supported from `0.5+` so we should be good to drop the fallback to notify. I've adjusted the error wording and some of the highlights. If you're happy with it, we can merge! Sorry for the delay on this one

Saghen commented:
Amazing work on this, thanks again!

OXY2DEV commented:
Happy to help!
